piptools/scripts/compile.py

```diff
@@ -477,14 +494,29 @@ def cli(
         find_links=repository.finder.find_links,
         emit_find_links=emit_find_links,
     )
+    markers = {key_from_ireq(ireq): ireq.markers for ireq in constraints
+               if ireq.markers}
     writer.write(
         results=results,
         unsafe_requirements=resolver.unsafe_constraints,
-        markers={
-            key_from_ireq(ireq): ireq.markers for ireq in constraints if ireq.markers
-        },
+        markers=markers,
         hashes=hashes,
     )
 
+    if split_wheels:
+        # append `_wheels` to wheels requirement output
+        name, ext = os.path.splitext(output_file.name)
```

Comment:
How about something like this.  The function could be nested, which would avoid the need to pass in writer and resolver and generate_hashes.

```python
def write(writer, resolver, results, generate_hashes, output_file):
    if generate_hashes:
        try:
            hashes = resolver.resolve_hashes(results)
        except PipToolsError as e:
            log.error(str(e))
            sys.exit(2)
    else:
        hashes = None
    writer.dst_file = output_file
    writer.write(
        results=results,
        unsafe_requirements=resolver.unsafe_constraints,
        markers=markers,
        hashes=hashes,
    )

if split_wheels:
    results_wheels = set([result for result in results
                                  if result.is_wheel])
    results -= results_wheels
    name, ext = os.path.splitext(output_file.name)
    wheels_output = click.open_file(
            name + WHEELS_REQUIREMENTS_SUFFIX + ext, "w+b", atomic=True,
            lazy=True)
    ctx.call_on_close(safecall(wheels_output.close_intelligently))
    write(writer, resolver, results_wheels, generate_hashes, wheels_output)

write(writer, resolver, results, generate_hashes, output_file)

```

Reply:
right, updated, thanks Chad!